src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java

```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
+
+        String lowerBoundModeValue = Objects.toString(lowerBound.get(PARAM_NAME_LOWER_BOUND_MODE), "");
+        LowerBound.Mode mode = LowerBound.Mode.fromString(lowerBoundModeValue);
+        float minScore = extractAndValidateMinScore(lowerBound);
+
+        return ImmutablePair.of(mode, minScore);
+    }
+
+    private float extractAndValidateMinScore(Map<String, Object> lowerBound) {
+        Object minScoreObj = lowerBound.get(PARAM_NAME_LOWER_BOUND_MIN_SCORE);
+        if (minScoreObj == null) {
+            return LowerBound.DEFAULT_LOWER_BOUND_SCORE;
+        }
+        try {
+            float minScore = LowerBound.DEFAULT_LOWER_BOUND_SCORE;
+            if (Objects.nonNull(lowerBound.get(PARAM_NAME_LOWER_BOUND_MIN_SCORE))) {
+                minScore = Float.parseFloat(String.valueOf(lowerBound.get(PARAM_NAME_LOWER_BOUND_MIN_SCORE)));
+            }
+            Validate.isTrue(
+                minScore >= LowerBound.MIN_LOWER_BOUND_SCORE && minScore <= LowerBound.MAX_LOWER_BOUND_SCORE,
+                "min_score must be a valid finite number between %f and %f",
+                LowerBound.MIN_LOWER_BOUND_SCORE,
+                LowerBound.MAX_LOWER_BOUND_SCORE
+            );
+            return minScore;
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("invalid format for min_score: must be a valid float value", e);
+        }
     }
 
     /**
      * Result class to hold min and max scores for each sub query
      */
     @AllArgsConstructor
     @Getter
-    private class MinMaxScores {
+    private static class MinMaxScores {
         float[] minScoresPerSubquery;
         float[] maxScoresPerSubquery;
     }
+
+    /**
+     * Result class to hold lower bound for each sub query
+     */
+    @Getter
+    public static class LowerBound {
+        static final float MIN_LOWER_BOUND_SCORE = -10_000f;
+        static final float MAX_LOWER_BOUND_SCORE = 10_000f;
+        static final float DEFAULT_LOWER_BOUND_SCORE = 0.0f;
+
+        private final boolean enabled;
+        private final Mode mode;
+        private final float minScore;
+
+        LowerBound() {
+            this(false, Mode.DEFAULT, DEFAULT_LOWER_BOUND_SCORE);
+        }
+
+        LowerBound(boolean enabled, Mode mode, float minScore) {
+            this.enabled = enabled;
+            this.mode = mode;
+            this.minScore = minScore;
+        }
+
+        /**
+         * Enum for normalization mode
+         */
+        protected enum Mode {
+            APPLY {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // if we apply the lower bound this mean we use actual score in case it's less then the lower bound min score
+                    // same applied to case when actual max_score is less than lower bound min score
+                    if (maxScore < lowerBoundScore || score < lowerBoundScore) {
```

Comment:
Why are we checking `maxScore` here? Shouldn't it be `minScore` that needs to be compared with `lowerBoundScore`

Reply:
this is an edge case I bumped into while working on implementation. Imagine most of your docs are not very relevant to the query and the lower bound min_score is relatively high - this is the scenario we're catching here. We effectively falling back to traditional min-max for the lack of better option.

Reply:
This should be common case, minScore should be lower then the lowerBound score in general. In this case we mainly case how actual score compares with lowerBound score, not minScore with lowerBound score. In case doc score is >= then lowerBound score we use lowerBound as min score:

```(score - lowerBoundScore) / (maxScore - lowerBoundScore);```

in case doc score < lowerBound score then we keep raw min-max formula:

```(score - minScore) / (maxScore - minScore)```